### PR TITLE
[BEAM-3744] Python PubSub API Fixes and Tests

### PR DESCRIPTION
**Pull request by udim**
Fixes attributes and adds an integration test for Dataflow.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

File: sdks/python/apache_beam/io/gcp/pubsub.py [hunk 1]
```diff
@@ -155,45 +175,94 @@ def to_runner_api_parameter(self, context):
...
+  def to_proto(element):
+    if not isinstance(element, PubsubMessage):
+      raise TypeError('Unexpected element. Type: %s (expected: PubsubMessage), '
+                      'value: %r' % (type(element), element))
+    return element._to_proto()
+
+  def expand(self, pcoll):
+    if self.with_attributes:
+      pcoll = pcoll | 'ToProtobuf' >> Map(self.to_proto)
+      pcoll.element_type = six.binary_type
```
Comment (charlesccychen) [inline sdks/python/apache_beam/io/gcp/pubsub.py]: Is this valid?  Shouldn't the output type be some protobuf type?

**Comment by charlesccychen** [comment]
It also looks like there is a test failure: https://builds.apache.org/job/beam_PreCommit_Python_Commit/421/testReport/junit/apache_beam.io.gcp.pubsub_test/TestPubsubMessage/test_proto_conversion/

File: sdks/python/apache_beam/io/gcp/pubsub.py (hunk 1 above)
Comment (udim) [reply sdks/python/apache_beam/io/gcp/pubsub.py]: After some investigation I believe that this is correct.
The output of this mapping is a string (uses the proto object's `SerializeToString()`).
On Dataflow, this string is passed to a runner harness written in Java, so it must be a serialized protobuf to be understood.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 2]
```diff
@@ -841,11 +841,12 @@ def run_Read(self, transform_node):
                           transform.source.id_label)
       if transform.source.with_attributes:
         # Setting this property signals Dataflow runner to return full
-        # PubsubMessages instead of just the payload.
+        # PubsubMessages instead of just data.
         step.add_property(PropertyNames.PUBSUB_SERIALIZED_ATTRIBUTES_FN, '')
       if transform.source.timestamp_attribute is not None:
         step.add_property(PropertyNames.PUBSUB_TIMESTAMP_ATTRIBUTE,
                           transform.source.timestamp_attribute)
+      logging.info('pubsub source')
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: debugging comment?

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 3]
```diff
@@ -841,11 +841,12 @@ def run_Read(self, transform_node):
                           transform.source.id_label)
       if transform.source.with_attributes:
         # Setting this property signals Dataflow runner to return full
-        # PubsubMessages instead of just the payload.
+        # PubsubMessages instead of just data.
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: "the payload data" for clarity?  This is the official doc for PubsubMessage: https://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py (hunk 2 above)
Comment (udim) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: Oops. Removed

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py (hunk 3 above)
Comment (udim) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: Fixed

**Comment by charlesccychen** [comment]
Thank you! This LGTM.

**Comment by aaltay** [comment]
Thank you! I appreciate this.

**Pull request merged**